File: backend/app/schemas/excel.py
"""Excel upload and preview schemas."""

from typing import Any, Optional

from pydantic import BaseModel


class ValidationWarning(BaseModel):
    level: str  # "warning" | "error" | "info"
    field: str
    message: str


class UploadResponse(BaseModel):
    upload_id: str
    original_filename: str
    file_size_bytes: int
    sheet_names: list[str]
    detected_project_type: Optional[str] = None
    detected_project_name: Optional[str] = None
    detection_confidence: Optional[float] = None
    validation_warnings: list[ValidationWarning] = []


class SheetInfo(BaseModel):
    name: str
    rows: int
    columns: int


class SheetData(BaseModel):
    name: str
    columns: list[str]
    rows: list[dict[str, Any]]
    total_rows: int
    page: int
    page_size: int


class SingleValuesResponse(BaseModel):
    fields: dict[str, Any]


class DetectResult(BaseModel):
    project_type: Optional[str] = None
    project_name: Optional[str] = None
    confidence: Optional[float] = None

File: frontend/src/api/excel.ts
import client from './client'

export interface ValidationWarning {
  level: 'warning' | 'error' | 'info'
  field: string
  message: string
}

export interface UploadResult {
  upload_id: string
  original_filename: string
  file_size_bytes: number
  sheet_names: string[]
  detected_project_type: string | null
  detected_project_name: string | null
  detection_confidence: number | null
  validation_warnings: ValidationWarning[]
}

export interface SheetInfo {
  name: string
  rows: number
  columns: number
}

export interface SheetData {
  name: string
  columns: string[]
  rows: Record<string, any>[]
  total_rows: number
  page: number
  page_size: number
}

export const excelApi = {
  async upload(file: File): Promise<UploadResult> {
    const form = new FormData()
    form.append('file', file)
    const { data } = await client.post('/excel/upload', form)
    return data.data
  },

  async getUpload(uploadId: string): Promise<UploadResult> {
    const { data } = await client.get(`/excel/${uploadId}`)
    return data.data
  },

  async getSheets(uploadId: string): Promise<SheetInfo[]> {
    const { data } = await client.get(`/excel/${uploadId}/sheets`)
    return data.data
  },

  async getSheetData(
    uploadId: string,
    sheetName: string,
    page = 1,
    pageSize = 50,
  ): Promise<SheetData> {
    const { data } = await client.get(
      `/excel/${uploadId}/sheets/${encodeURIComponent(sheetName)}`,
      { params: { page, page_size: pageSize } },
    )
    return data.data
  },

  async getSingleValues(uploadId: string): Promise<Record<string, any>> {
    const { data } = await client.get(`/excel/${uploadId}/single-values`)
    return data.data.fields
  },

  async detect(uploadId: string) {
    const { data } = await client.get(`/excel/${uploadId}/detect`)
    return data.data
  },
}
